### ISSUE TITLE
refresh on aws.s3.BucketObject source shows unnecessary update

### ISSUE THREAD

**Issue by clstokes** [pulumi/pulumi-aws#1521]
Provisioning a `aws.s3.BucketObject` using `source: FileAsset("file")` results in a confusing diff on the next `pulumi refresh`.

Unfortunately `ignoreChanges: ["source"]` does not seem to ignore the change from the refresh as it still shows up as a diff.

## Steps to reproduce

1. `git clone https://gist.github.com/a79ae702a7b79fc71a648fbae365bfc6`
2. `pulumi up -y --skip-preview`
3. `pulumi refresh`
    - observe diff

### Expected

No diff

### Actual

```
% pulumi up -y --skip-preview 
Updating (dev)

View Live: https://app.pulumi.com/clstokes/p-ts-s3-bucketobject-refresh/dev/updates/10

     Type                    Name                              Status      
 +   pulumi:pulumi:Stack     p-ts-s3-bucketobject-refresh-dev  created     
 +   ├─ aws:s3:Bucket        my-bucket                         created     
 +   └─ aws:s3:BucketObject  my-obj                            created     
 
Resources:
    + 3 created

Duration: 2m56s

% pulumi refresh
Previewing refresh (dev)

View Live: https://app.pulumi.com/clstokes/p-ts-s3-bucketobject-refresh/dev/previews/a76b793f-4446-4415-9946-1cb1c5e8e00e

     Type                    Name                              Plan       Info
     pulumi:pulumi:Stack     p-ts-s3-bucketobject-refresh-dev             
 ~   ├─ aws:s3:BucketObject  my-obj                            update     [diff: ~source]
     └─ aws:s3:Bucket        my-bucket                                    
 
Resources:
    ~ 1 to update
    2 unchanged

Do you want to perform this refresh?
Do you want to perform this refresh? details
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::p-ts-s3-bucketobject-refresh::pulumi:pulumi:Stack::p-ts-s3-bucketobject-refresh-dev]
    ~ aws:s3/bucketObject:BucketObject: (update)
        [id=my-obj]
        [urn=urn:pulumi:dev::p-ts-s3-bucketobject-refresh::aws:s3/bucketObject:BucketObject::my-obj]
        [provider=urn:pulumi:dev::p-ts-s3-bucketobject-refresh::pulumi:providers:aws::default_4_7_0::5be630b8-4d90-4b86-b07d-b64e23da802b]
        --outputs:--
      - source                   : asset(file:56f4783) { Pulumi.yaml }
      + source                   : "/var/folders/z8/fbg1lmkx4pb28gr3m6_1v_xh0000gn/T/pulumi-asset-56f47834e39fbfba1a5e66a62065534ab285b664aa9f5f905a7ab092fccc0d91"

```

**Comment by bdenkmann** [comment]
This results in an error after we **refresh** and attempt to **up** again: 

`error: Running program 'C:\docusign_source\Nexus\CloudResources\aws\Nexus_Aws\doculake_main\bin\Debug\netcoreapp3.1\doculake_main.dll' failed with an unhandled exception:
    System.InvalidOperationException: Expected Pulumi.AssetOrArchive but got System.String deserializing Pulumi.Aws.S3.BucketObject.source
       at Pulumi.Serialization.Converter.ConvertObject(String context, Object val, Type targetType)
       at Pulumi.Serialization.Converter.ConvertValue(String context, Value value, Type targetType, ImmutableHashSet`1 resources)
       at Pulumi.Deployment.CompleteResourceAsync(Resource resource, Boolean remote, Func`2 newDependency, ResourceArgs args, ResourceOptions options, ImmutableDictionary`2 completionSources)
       at Pulumi.Deployment.Runner.<>c__DisplayClass9_0.<<WhileRunningAsync>g__HandleCompletion|0>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at Pulumi.Deployment.Runner.WhileRunningAsync()`

**Comment by bdenkmann** [comment]
The only way we have found to work around this is to delete the assets in S3 manually, **refresh**, and then **up**. We tried using **IgnoreChanges** but couldn't get the diff to be ignored using this method.

**Comment by lukehoban** [comment]
Also tracked in https://github.com/pulumi/pulumi/issues/6235.

**Comment by pgavlin** [comment]
We've addressed the crash in order to unblock users who are hitting this issue: https://github.com/pulumi/pulumi/pull/7579. These changes will be in the next release of the Pulumi CLI. I'm going to leave this issue open to track fixing the 'unexpected update' aspect of the problem.

**Comment by phillipedwards** [comment]
With the changes in in [pulumi/pulumi#7579](https://github.com/pulumi/pulumi/pull/7579), the S3 Bucket Object errors are resolved. That said, a very similar error is occurring for C# Pulumi programs that are created lambda functions with FileArchives. This error makes me think this is related to the fix released with Pulumi v3.8.0.

![image](https://user-images.githubusercontent.com/25461821/127074064-541dcb94-11d6-4d59-a820-21e87cfcb569.png)

This can be seen by executing this [csharp pulumi example](https://github.com/pulumi/examples/tree/master/aws-cs-lambda). If this would be better suited in a new issue, I can created it as such.

**Comment by pgavlin** [comment]
Hm, that's surprising--we added code to the .NET SDK to allow this. @justinvp does NuGet have an equivalent to `yarn.lock` that we can use to see exactly which version of the .NET SDK is in use?

**Comment by phillipedwards** [comment]
@pgavlin this may be a non-issue. Updating .net sdk to latest (4.14.0) appears to behave as expected with no error.

**Comment by pgavlin** [comment]
> @pgavlin this may be a non-issue. Updating .net sdk to latest (4.14.0) appears to behave as expected with no error.

Phew. Glad to hear that--thanks for following up!

**Comment by lukehoban** [comment]
This is an issue ing `pulumi/pulumi`, and is being tracked in https://github.com/pulumi/pulumi/issues/6235, so I'll close this one out as a duplicate.
